gmp-droid.cpp

```diff
@@ -239,97 +231,97 @@ class DroidVideoDecoder : public GMPVideoDecoder
   void SubmitBufferThread (DroidMediaCodecData cdata,
       DroidMediaBufferCallbacks cb)
   {
-    m_drain_lock->Acquire ();
+    m_codec_lock->Acquire ();
+
     if (m_draining || (!m_codec && !CreateCodec ())) {
       LOG (ERROR, "Buffer submitted while draining");
       cb.unref (cb.data);
-      m_drain_lock->Release ();
-      return;
+      goto out_unlock;
     }
-    m_drain_lock->Release ();
 
     if (m_resetting) {
       LOG (INFO, "Buffer submitted while resetting");
-      return;
+      goto out_unlock;
     }
 
+    m_codec_lock->Release ();
+
     // This blocks when the input Source is full
     droid_media_codec_queue (m_codec, &cdata, &cb);
 
-    m_drain_lock->Acquire ();
-    if (!m_draining && m_callback && g_platform_api) {
-      g_platform_api->runonmainthread (WrapTask (m_callback,
-              &GMPVideoDecoderCallback::InputDataExhausted));
+    m_codec_lock->Acquire ();
+
+    if (g_platform_api) {
+      g_platform_api->runonmainthread (WrapTask (this,
+              &DroidVideoDecoder::InputDataExhausted_m));
     }
-    m_drain_lock->Release ();
+out_unlock:
+    m_codec_lock->Release ();
   }
 
-  virtual void Reset ()
-  {
-    m_stop_lock->Acquire ();
-    if (m_resetting) {
-      m_stop_lock->Release ();
-      return;
+  void InputDataExhausted_m() {
+    m_codec_lock->Acquire ();
+    if (!m_draining && m_callback) {
+      m_callback->InputDataExhausted();
     }
+    m_codec_lock->Release ();
+  }
 
-    m_resetting = true;
+  virtual void Reset ()
+  {
+    m_codec_lock->Acquire ();
 
-    if (m_processing) {
-      // Reset() will be called from DataAvailable() later
-      LOG (INFO, "Reset while m_processing");
-      m_stop_lock->Release ();
-      return;
+    if (!m_resetting && m_submit_thread) {
+      m_submit_thread->Post (WrapTask (this,
+              &DroidVideoDecoder::ResetCodec));
     }
 
-    if (g_platform_api) {
-      // Reset() was called. Execute it on main thread
-      g_platform_api->runonmainthread (WrapTask (this,
-              &DroidVideoDecoder::Reset_m));
-    }
-    m_stop_lock->Release ();
+    m_codec_lock->Release();
+
   }
 
   virtual void Drain ()
```

Comment:
What thread is this called from? From what I can follow other references to m_codec are confined to the submit thread in which case there's no need to protect access to it, but here I'm wondering if it's possible for the codec to be destroyed from the the submit thread mid-drain. If you can't call droid_media_codec within a lock is it possible acquire a local reference to codec while locked and then unlock, drain and release the local reference?

Access to m_draining is protected elsewhere too but not at the top here, and then it gets cleared after the drain which looks like it could turn into a timing issue.